### PR TITLE
Add Initialize Git CheckBox button to Create Project dialogue

### PR DESCRIPTION
**Pull request by Salvakiya**
This add LibGit2Sharp dependency to the project (which I plan to implement a git submodule frontend for managing addons using Godot-Manager)

There is a caveat. I had to copy the resulting git2-xxxxxxx.dll file from the ".mono\temp\bin\Debug\lib\win32\x64" folder and place it alongside the project for it to run properly. This file is a dependency of LibGit2Sharp. I did not want to commit the dll file itself. I am relatively new to C# and am unsure how to fix that issue. On export Godot and C# should include this dependency somehow.

### CODE REVIEW

File: Scripts/components/Dialogs/CreateProject.cs [hunk 1]
```diff
@@ -7,198 +7,202 @@ public class CreateProject : ReferenceRect
 {
 
 #region Signals
-    [Signal]
-    public delegate void project_created(ProjectFile projFile);
+	[Signal]
+	public delegate void project_created(ProjectFile projFile);
```
Comment (aaronfranke) [inline Scripts/components/Dialogs/CreateProject.cs:11]: The C# style guide says to use spaces, not tabs. You are replacing spaces with tabs here.

Comment (eumario) [reply Scripts/components/Dialogs/CreateProject.cs:11]: Plus, I have made changes to CreateProject.cs/tscn and NewProject to support Godot 4 project files, so these need to be resolved.

**Comment by eumario** [comment]
The reason why you need to manually copy of the DLL file is because you do not have a reference to LibGit2Sharp.NativeBinaries in the csproj file.  This would ensure that when building the project, the C native library that is needed for the specified platform that the project is being built for, has the correct Shared library.  Windows is a DLL, Linux is a SO, and Mac is a DYNLIB.  Adding the reference to the NativeBinaries project, will ensure these binaries are properly copied when compiling the project.

File: Scripts/components/Dialogs/CreateProject.cs (hunk 1 above)
Comment (Salvakiya) [reply Scripts/components/Dialogs/CreateProject.cs:11]: I didnt even realise lol. I will make the changes. Thanks!

**Comment by Salvakiya** [comment]
> The reason why you need to manually copy of the DLL file is because you do not have a reference to LibGit2Sharp.NativeBinaries in the csproj file. This would ensure that when building the project, the C native library that is needed for the specified platform that the project is being built for, has the correct Shared library. Windows is a DLL, Linux is a SO, and Mac is a DYNLIB. Adding the reference to the NativeBinaries project, will ensure these binaries are properly copied when compiling the project.

ahh okay. I thought it must have been something simple like that.

**Comment by Salvakiya** [comment]
I could not figure out the LibGit2Sharp and LibGit2Sharp.NativeBinary issue. Even after adding it as a separate line to the csproj file. when testing and on export it still failed to include the appropriate dll file.

I looked up another way to run git commands using C# and am currently doing that for now. Also it copy's over a gitingore if you initialize a git repo on a new project.

ALSO: I added an ErrorWindow which is a child of SceneManager. This is an easy way to display error messages to the user which should help with debugging. Please let me know if you would like this placed somewhere else?

**Comment by eumario** [comment]
The second method in which to use Git, will require you to know where Git is, as sometimes it may not be found correctly.  The best test for that, would be to look at how I do chmod and xattr for Linux and Mac OS respectively.  This will need to be done on Windows as well, which can be done with the where command, EG: "where git" will return the location of where Git is installed.

It's always best to ensure you have the full path to the Executable, when running commands, as it leaves nothing for chance of not being there.

And there should be error handlers for when the command is not found.  chmod is installed on all Linux systems, same with xattr is installed on all Mac OS systems, so there's no worry about them not being there.  But git is not installed by default, so it would be best to get the full location to Git, and error out if it can't find it, offer an option to the user to define where their git install is, or instruct them on how to install git.

As for the ErrorWindow.  I am guessing your wanting a window dialog that just allows for multi-line text in it?  If not, there is AppDialog.MessageDialog, AppDialog.YesNo and AppDialog.YesNoCancel which can be used to display a message to the user, as well as ask questions.

File: libs/managers/GitTools.cs [hunk 2]
```diff
@@ -3,14 +3,40 @@
...
+using System.Runtime.InteropServices;
+using Godot;
+using Godot.Collections;
 namespace GitTools
 {
-    public class GitRunner
+    class Util{
+        public static String GetGit()
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Comment (Salvakiya) [inline libs/managers/GitTools.cs:14]: I assume this is more what we are looking for

File: libs/managers/GitTools.cs [hunk 3]
```diff
@@ -3,14 +3,40 @@
...
+            {
+                var output = new Godot.Collections.Array();
+                int exit_code = OS.Execute("where", new string[] { "git" }, true, output);
+                if (exit_code == 0)
+                {
+                    return (output[0] as string).StripEdges();
+                }
+            }
+            else// if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux)) //Which can be used on linux and mac?
```
Comment (Salvakiya) [inline libs/managers/GitTools.cs:23]: I understand that linux and mac use the same command? and this should work?

**Comment by Salvakiya** [comment]
still not near done but I want to make sure I am working on this a little every week. Love the tool!

File: libs/managers/GitTools.cs (hunk 2 above)
Comment (eumario) [reply libs/managers/GitTools.cs:14]: The way that I determine it in Godot is with Compile Time flags, such as:
#if GODOT_WINDOWS || GODOT_UWP  // Windows
#elif GODOT_LINUXBSD || GODOT_X11 // Linux
#elif GODOT_MACOS || GODOT_OSX // Mac OS
#endif

This way, only the code between the true if statement will be put in, based upon the platform it is being compiled for.

File: libs/managers/GitTools.cs (hunk 3 above)
Comment (eumario) [reply libs/managers/GitTools.cs:23]: This is correct, Windows = where, Linux and Mac OS = which.

**Comment by eumario** [comment]
Take your time, it's not something that you absolutely have to do right this moment.  As you can see from my own commits, I take my time getting stuff correct. ;-)  Or at least try to.

File: libs/managers/GitTools.cs (hunk 2 above)
Comment (Salvakiya) [reply libs/managers/GitTools.cs:14]: Ahh I like that.. silly me. Still new to C# but have used C before.

**Comment by eumario** [comment]
Since no progress has been made on this Pull request since April, I am closing the PR, as a Major Re-write will be coming soon, that will re-work a lot of things, and may provide an easier path to working with Git in the future.

**Pull request closed without merging**